### ISSUE TITLE
Support stack probing for arm64 (aarch64)

### ISSUE THREAD

**Issue by tmandry** [rust-lang/rust#77071]
Background: #43241

This issue is to track implementation of stack probing on arm64. Most of the work necessary is in in LLVM, but this is a convenient place to track progress and final support in the Rust compiler.

Implementation: @cuviper has suggested using LLVM's inline-asm support for stack probing: https://github.com/rust-lang/rust/issues/43241#issuecomment-594099684. It does not yet support arm64.

External trackers:
- https://linaro.atlassian.net/browse/LLVM-618

**Comment by tmandry** [comment]
Hi @raw-bin, what's the status of this? As I recall this work was planned to land as part of LLVM 12, but I haven't seen updates on the Linaro tracker yet.

**Comment by raw-bin** [comment]
Hi Tyler.

I believe there has been progress but let me get the definitive and circle
back ASAP.

Cheers,
Robin

**Comment by raw-bin** [comment]
Hi @tmandry, folks.

Apologies for the delay.

There has been good progress.

Please note the following patches that have been posted by Linaro for review:

https://reviews.llvm.org/D96004
https://reviews.llvm.org/D96005
https://reviews.llvm.org/D96006
https://reviews.llvm.org/D96007

At present Linaro is blocking on review.

If possible, please do chime in.

I will provide further updates as needed.

Cheers

**Comment by tmandry** [comment]
Great, thanks for the update! Let me ask around to see if I can help get those patches reviewed.

**Comment by joshtriplett** [comment]
What's the current state of this? What needs to happen to get working stack probes on aarch64?

**Comment by pnkfelix** [comment]
Visited for T-compiler steering backlog bonanza.

Seems like 3 out of the 4 of the LLVM patches linked above are still undergoing review. So, tagging this as unimplemented. (And also blocked, since its waiting on LLVM review capacity that we really can't do much about ourselves.)

@rustbot label: S-tracking-unimplemented S-blocked

**Comment by jacobbramley** [comment]
Arm's LLVM team have now picked this up, and have told us that the work should be completed by September at the latest (LLVM 17).

**Comment by tstellar** [comment]
The deadline to get changes into LLVM17 is the end of July, so September would be too late.

**Comment by jacobbramley** [comment]
Yes, sorry, I wasn't very clear. They're hoping to get this reviewed and merged in time for the LLVM 17 release branch, which means by the end of July on `main`. There should therefore be something for us to see and test somewhat earlier than September, but (according to [LLVM's release process documentation](https://llvm.org/docs/HowToReleaseLLVM.html))`17.0.0-final` would take at least an additional six weeks, and of course it's possible that issues are found in that testing time.

**Comment by jonathanpallant** [comment]
Does anyone know if these changes made it in?

**Comment by jacobbramley** [comment]
They didn't! I'd been tracking it, and all seemed to be going well until recently, so I've been digging a bit to work out what's happened.

It seems that the work was progressing as planned, but hit validation issues, so it missed the LLVM17 code freeze, sadly. It was being tracked in [Linaro's LLVM-618](https://linaro.atlassian.net/browse/LLVM-618), but that doesn't describe the exact problem. I'm still hoping to have it landed on `main` somewhat before the LLVM18 code freeze, and I've stressed its importance.

I apologise for the further delay!

**Comment by jacobbramley** [comment]
This is in progress here: llvm/llvm-project#66524 (moved from [Phabricator](https://reviews.llvm.org/D158084))

**Comment by slanterns** [comment]
It has been merged. Hopefully we can get it after upgrading to LLVM18.

**Comment by cuviper** [comment]
Great! If someone wants to try it, you can add `stack_probes: StackProbeType::Inline` to the options here:
https://github.com/rust-lang/rust/blob/b10cfcd65fd7f7b1ab9beb34798b2108de003452/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_gnu.rs#L9
... and build the compiler with your own LLVM 18 / main.

(The inline option should be a no-op on versions where the target didn't support it.)

Then `tests/ui/abi/stack-probes.rs` can be updated for a functional test, but even just the normal rustc bootstrap process will probably exercise probing enough to see if there are any rough edges.

**Comment by cuviper** [comment]
I went ahead and tested it myself and found no issues => #118491 flips the switch!

**Comment by erikschul** [comment]
The docs claim that `Tier 1 targets can be thought of as "guaranteed to work".`
perhaps the target should be removed from tier 1 if it hasn't worked for 3 years?

Currently I've spent hours trying to cross-compile from aarch64 to x86_64.
`cargo build --target=x86_64-unknown-linux-gnu` raises this error:
```
cc: error: unrecognized command-line option '-m64'
```
where `/usr/bin/x86_64-linux-gnu-gcc  -E -march=help -xc /dev/null` suggests that the flag should be `-mx86-64`.
Is there an environment variable that can override this specific flag?

**Comment by cuviper** [comment]
@erikschul Cross-compilation is a completely different thing -- please open a separate issue.

**Comment by adamgemmell** [comment]
I've done a top 20k crater run on cuviper's https://github.com/rust-lang/rust/pull/118491 on an aarch64 machine. I used Rust e51e98dde6a60637b6a71b8105245b629ac3fe77 and LLVM c09e6905567a6b546bb2fd9e863511a8fb939b19 and didn't find any non-spurious issues.

**Comment by jacobbramley** [comment]
Is this issue now complete, or do we need anything else?

**Comment by jonathanpallant** [comment]
What about the `1` footnote on https://doc.rust-lang.org/nightly/rustc/platform-support.html?

**Comment by cuviper** [comment]
I think we're done! Aarch64 stack probes are tested in CI (e.g. [here](https://github.com/rust-lang-ci/rust/actions/runs/8190263137/job/22396848497#step:27:9505) and [here](https://github.com/rust-lang-ci/rust/actions/runs/8190263137/job/22396848497#step:27:10589)), and #122094 removed that footnote.

If there are fundamental problems discovered, we can reopen this, otherwise please open new issues for any bugs.
